feat(ZNTA-2318): Editor Review / Reject Translation

efloden opened the pull request:
JIRA issue URL: https://zanata.atlassian.net/browse/ZNTA-2318

Modification of the RejectTranslationModal frontend container and backend rest API in progress to implement this feature.

## QA
The alpha editor now has the approve and reject options for each text flow.
Clicking reject or selecting it with the key combination ( CTRL + ⇧ + S R ) should open a modal with criteria and priority dropdown fields, and a comment text field.
The priority field selection has not been implemented in the backend as the details around this option are still being discussed #672 .
The comment is also not required to submit a rejection.
Behavior when there are no criteria to select is also still being determined, this PR can be tested with existing criteria (added through frontend admin panel -> review criteria)
After adding a review comment it should be observable in the GWT editors translation history.
![rejecting](https://user-images.githubusercontent.com/7971187/35902521-3a41d55a-0c27-11e8-98c9-c8ff0a7709e0.png)

## Checklist

- JIRA link
- Check target branch
- Make sure all commit statuses are green or otherwise documented reasons to ignore
- QA needs to evaluate against the JIRA ticket
- Changed files and commits make sense for this PR

See [Zanata Development Guidelines](https://github.com/zanata/zanata-platform/wiki/Development-Guidelines) more for information.

----
*This template can be updated in .github/PULL_REQUEST_TEMPLATE.md*

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/zanata/zanata-platform/688)
<!-- Reviewable:end -->

huangp commented:
Reviewed 4 of 10 files at r2, 1 of 4 files at r3, 7 of 14 files at r4, 10 of 10 files at r5.
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[server/services/src/main/java/org/zanata/rest/editor/service/resource/TranslationReviewResource.java, line 22 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4Etbzukni8IJi08Peo:-L4Etbzukni8IJi08Pep:b-fu0gxa) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/services/src/main/java/org/zanata/rest/editor/service/resource/TranslationReviewResource.java#L22)):*
> ```Java
> 
>     /**
>      * Update/insert translation
> ```

comment needs update

---

*[server/zanata-frontend/src/app/editor/actions/review-trans-actions.js, line 17 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4EudVlkjqV-SRUqMUq:-L4EudVlkjqV-SRUqMUr:b-ckfxji) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/actions/review-trans-actions.js#L17)):*
> ```JavaScript
> /**
>  * Perform a save with the given info, and recursively start next save if
>  * one has queued when the save finishes.
> ```

Comment needs update

---

*[server/zanata-frontend/src/app/editor/actions/review-trans-actions.js, line 19 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4EuAet-iXWAzgDSVtZ:-L4EuAeucKCawdxYj68V:b-uq0mg8) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/actions/review-trans-actions.js#L19)):*
> ```JavaScript
>  * one has queued when the save finishes.
>  */
> export function addNewReview (review, localeId) {
> ```

I think we should rename the function to 'rejectTranslation'. Since the status is always save as Rejected

---

*[server/zanata-frontend/src/app/editor/components/TransUnit/index.js, line 88 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4EvAXi-VtDa_axIMpP:-L4EvAXi-VtDa_axIMpQ:b-6dc4a0) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/components/TransUnit/index.js#L88)):*
> ```JavaScript
>     // Only Fetch the criteria for this translation we intend to reject
>     if (!this.state.showRejectModal) {
>       this.props.fetchAllCriteria()
> ```

We could fetch all criteria at the beginning of loading the editor, and cache the result in redux store's state. This saves us from loading the same list over and over again.

---

*[server/zanata-frontend/src/app/editor/components/TransUnit/index.js, line 141 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4Ew47gpj5hPB3ClMpe:-L4Ew47gpj5hPB3ClMpf:bcqv1xh) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/components/TransUnit/index.js#L141)):*
> ```JavaScript
>     const rejectTranslationModal = this.state.showRejectModal
>       ? <RejectTranslationModal
>         show={this.state.showRejectModal}
> ```

we already have a show prop here, do we still need the outer condition check?

---

*[server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/CriteriaDropdown.js, line 8 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4EwL3OBD9uWP-L__3o:-L4EwL3OBD9uWP-L__3p:b-5woj9k) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/CriteriaDropdown.js#L8)):*
> ```JavaScript
> export const MINOR = 'Minor'
> export const MAJOR = 'Major'
> export const CRITICAL = 'Critical'
> ```

These also appear in components/TransUnit/index.js. We could move it in a file under util and share it.

---

*[server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/CriteriaDropdown.js, line 9 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4EwHb75MkfoIpMGExz:-L4EwHb75MkfoIpMGEy-:bgsv7yp) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/CriteriaDropdown.js#L9)):*
> ```JavaScript
> export const MAJOR = 'Major'
> export const CRITICAL = 'Critical'
> /* eslint-disable max-len */
> ```

do we still need this?

---

*[server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js, line 12 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4ExdN4zDJaP6cUUiS6:-L4ExdN4zDJaP6cUUiS7:b-c278r0) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js#L12)):*
> ```JavaScript
> import update from 'immutability-helper'
> 
>  /* eslint-disable max-len */
> ```

Still need this?

---

*[server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js, line 16 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4Exfvj8GWevN-JZzKH:-L4Exfvj8GWevN-JZzKI:b9xowsf) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js#L16)):*
> ```JavaScript
> export const MAJOR = 'Major'
> export const CRITICAL = 'Critical'
> export const priorities = [MINOR, MAJOR, CRITICAL]
> ```

move to util file

---

*[server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js, line 27 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4F-2U5rauo25s_8v1h:-L4F-2U5rauo25s_8v1i:b-f5krfy) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js#L27)):*
> ```JavaScript
>     transUnitID: PropTypes.number.isRequired,
>     revision: PropTypes.number.isRequired,
>     language: PropTypes.string.isRequired,
> ```

I would suggest to rename this to localeId. We all know what localeId is but language as string type could mean other things.

---

*[server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js, line 38 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4F-GlKsCn74hHmAy3h:-L4F-GlKsCn74hHmAy3i:baijqdu) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js#L38)):*
> ```JavaScript
>     review: {
>       id: 0,
>       revision: 1,
> ```

id and revision should not be part of the state. They are available from props.

---

*[server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js, line 61 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4Ey5IAw48OrPmKoH8n:-L4Ey5IAw48OrPmKoH8o:btfep2z) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js#L61)):*
> ```JavaScript
>   }
>   onPriorityChange = (event) => {
>     event.persist()
> ```

we could just introduce a local variable to hold event.target.innerText and use it in the setState callback. No need to use event.persist

---

*[server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js, line 71 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4EyD0Y2BKvKynMCjZ6:-L4EyD0Y2BKvKynMCjZ7:b-lteavi) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js#L71)):*
> ```JavaScript
>   }
>   onCriteriaChange = (event) => {
>     event.persist()
> ```

should not use event.persist

---

*[server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js, line 82 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4EyXqx24_a3YaBHtta:-L4EyXqx24_a3YaBHttb:b-lteavi) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js#L82)):*
> ```JavaScript
>   }
>   setReviewComment = (event) => {
>     event.persist()
> ```

should not use event.persist

---

*[server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js, line 89 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4EyjkXg35Q-QLMQrPK:-L4EyjkXg35Q-QLMQrPL:b-vfmfkt) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js#L89)):*
> ```JavaScript
>     }))
>   }
>   textState = () => {
> ```

this function can go into the util file (a function takes a priority and return a css class name)

---

*[server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js, line 109 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4F-TM_5xe-rF3cdqe9:-L4F-TM_5xe-rF3cdqeA:b-fmzbid) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js#L109)):*
> ```JavaScript
>       review
>     } = this.state
>     const saveTransReview = () => {
> ```

this function should be defined in the component. And the review parameter can be constructed from this.props and this.state

---

*[server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js, line 129 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4EzKrN5DkzIzoU0FId:-L4EzKrN5DkzIzoU0FIe:b-kzvqvs) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js#L129)):*
> ```JavaScript
>               onCriteriaChange={this.onCriteriaChange}
>               selectedCriteria={review.selectedCriteria} />
>             <PriorityDropdown
> ```

I don't think we need to present a priority dropdown to the user. Check with Kathryn and others to see whether reviewer is allowed to change the priority (after admin has created the criteria)

---

*[server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/PriorityDropdown.js, line 9 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4F-qa3m4E9CzxJNiCX:-L4F-qa3m4E9CzxJNiCY:bezr7fz) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/PriorityDropdown.js#L9)):*
> ```JavaScript
> export const MINOR = 'Minor'
> export const MAJOR = 'Major'
> export const CRITICAL = 'Critical'
> ```

Move to util file as the TODO says. But check with others to see if we even need to present the dropdown in editor.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/688)*
<!-- Sent from Reviewable.io -->

efloden commented:
Review status: 20 of 22 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[server/services/src/main/java/org/zanata/rest/editor/service/resource/TranslationReviewResource.java, line 22 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4Etbzukni8IJi08Peo:-L4JYWDxq2u8qNN6-bw4:bvnrpcm) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/services/src/main/java/org/zanata/rest/editor/service/resource/TranslationReviewResource.java#L22)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

comment needs update
</blockquote></details>

updated

---

*[server/zanata-frontend/src/app/editor/actions/review-trans-actions.js, line 17 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4EudVlkjqV-SRUqMUq:-L4JYXJEp5vVN7R-VVkq:bvnrpcm) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/actions/review-trans-actions.js#L17)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

Comment needs update
</blockquote></details>

updated

---

*[server/zanata-frontend/src/app/editor/actions/review-trans-actions.js, line 19 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4EuAet-iXWAzgDSVtZ:-L4F92NaOnw2C3rwk4QJ:b-bqw7cc) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/actions/review-trans-actions.js#L19)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

I think we should rename the function to 'rejectTranslation'. Since the status is always save as Rejected
</blockquote></details>

refactored 'addNewReview' to 'rejectTranslation'

---

*[server/zanata-frontend/src/app/editor/components/TransUnit/index.js, line 88 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4EvAXi-VtDa_axIMpP:-L4JXeZVBPfnx3Ffbglm:br60u1c) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/components/TransUnit/index.js#L88)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

We could fetch all criteria at the beginning of loading the editor, and cache the result in redux store's state. This saves us from loading the same list over and over again.
</blockquote></details>

Moved the fetchAllCriteria call up to Mainconent.js and passing it in to the TransUnit's so it only needs to be called once

---

*[server/zanata-frontend/src/app/editor/components/TransUnit/index.js, line 141 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4Ew47gpj5hPB3ClMpe:-L4F8kNsl46_HTBMPtt2:b-rqf2z8) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/components/TransUnit/index.js#L141)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

we already have a show prop here, do we still need the outer condition check?
</blockquote></details>

Rendering the modal and hiding it for each textflow seems to hurt performance quite a bit

---

*[server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/CriteriaDropdown.js, line 8 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4EwL3OBD9uWP-L__3o:-L4JRNW-9vtT39X8acGy:b-10q8hs) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/CriteriaDropdown.js#L8)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

These also appear in components/TransUnit/index.js. We could move it in a file under util and share it.
</blockquote></details>

importing these from a util file now

---

*[server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/CriteriaDropdown.js, line 9 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4EwHb75MkfoIpMGExz:-L4JD2M2iS0Ssfb9-rIM:b-4f4pze) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/CriteriaDropdown.js#L9)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

do we still need this?
</blockquote></details>

doesn't look like it, removed

---

*[server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js, line 12 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4ExdN4zDJaP6cUUiS6:-L4JLWpsqrjqA536-jew:b-th0y37) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js#L12)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

Still need this?
</blockquote></details>

removed, but had to make some single strings into multiline for placeholder and className props

---

*[server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js, line 16 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4Exfvj8GWevN-JZzKH:-L4JRWLp1JgGO9mcckfm:b-10q8hs) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js#L16)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

move to util file
</blockquote></details>

importing these from a util file now

---

*[server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js, line 27 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4F-2U5rauo25s_8v1h:-L4JDilXI7Cn-jRCOFM-:bx6bs0n) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js#L27)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

I would suggest to rename this to localeId. We all know what localeId is but language as string type could mean other things.
</blockquote></details>

renamed 'language' to 'localeId'

---

*[server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js, line 38 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4F-GlKsCn74hHmAy3h:-L4JFAVXpTyAe1YgwByF:b-x9tmgm) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js#L38)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

id and revision should not be part of the state. They are available from props.
</blockquote></details>

removed these from the state and passing them directly from props

---

*[server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js, line 61 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4Ey5IAw48OrPmKoH8n:-L4JGMMcX6QEZ4FPof2D:b-muoezh) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js#L61)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

we could just introduce a local variable to hold event.target.innerText and use it in the setState callback. No need to use event.persist
</blockquote></details>

using local variable rather than event.persist

---

*[server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js, line 71 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4EyD0Y2BKvKynMCjZ6:-L4JGQtXf3lXBkOfoyLo:b1w3n7r) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js#L71)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

should not use event.persist
</blockquote></details>

using local variable rather than event.persist

---

*[server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js, line 82 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4EyXqx24_a3YaBHtta:-L4JGVjulkeEQmdbdr9i:b-muoezh) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js#L82)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

should not use event.persist
</blockquote></details>

using local variable rather than event.persist

---

*[server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js, line 89 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4EyjkXg35Q-QLMQrPK:-L4JRXpdDMl70Qqkvne7:b-tivfef) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js#L89)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

this function can go into the util file (a function takes a priority and return a css class name)
</blockquote></details>

importing this function from a util file now

---

*[server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js, line 109 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4F-TM_5xe-rF3cdqe9:-L4JMD8INctqRi6hTeob:b-xbfnzw) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js#L109)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

this function should be defined in the component. And the review parameter can be constructed from this.props and this.state
</blockquote></details>

defined in the component, constructing the review param from this.props and this.state now

---

*[server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js, line 129 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4EzKrN5DkzIzoU0FId:-L4JTu5i4GaHGEOECAgl:bk0iitn) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js#L129)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

I don't think we need to present a priority dropdown to the user. Check with Kathryn and others to see whether reviewer is allowed to change the priority (after admin has created the criteria)
</blockquote></details>

discussion is on-going but for now it appears we want to give users the option: https://github.com/zanata/zanata-platform/pull/672

---

*[server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/PriorityDropdown.js, line 9 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4F-qa3m4E9CzxJNiCX:-L4JU-fM2w8-B7haUPPP:b-20eoah) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/PriorityDropdown.js#L9)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

Move to util file as the TODO says. But check with others to see if we even need to present the dropdown in editor.
</blockquote></details>

moved to util file

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/688)*
<!-- Sent from Reviewable.io -->

huangp commented:
Reviewed 1 of 2 files at r6, 10 of 11 files at r7.
Review status: 25 of 26 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[server/zanata-frontend/src/app/editor/components/TransUnit/index.js, line 141 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4Ew47gpj5hPB3ClMpe:-L4K6ykzaSlp2DcKosfp:b-b4q1lt) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/components/TransUnit/index.js#L141)):*
<details><summary><i>Previously, efloden (Earl Floden) wrote…</i></summary><blockquote>

Rendering the modal and hiding it for each textflow seems to hurt performance quite a bit
</blockquote></details>

oh I just realize we are rendering the modal component on every TransUnit. That's unnecessary. We should just render one RejectTranslationModal on the editor page (e.g. in mainContent). We can get selected transUnitID and current localeID from state.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/688)*
<!-- Sent from Reviewable.io -->

aeng commented:
@efloden build failed.

aeng commented:
Reviewed 1 of 11 files at r7, 7 of 9 files at r8.
Review status: 24 of 26 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[server/zanata-frontend/src/app/editor/components/TransUnitTranslationFooter.js, line 94 at r8](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4cgeZKi5lIm7U4WWPr:-L4cgeZLHHr3gz2elKxT:b-uum0ur) ([raw file](https://github.com/zanata/zanata-platform/blob/2508c8e86875b1baf135e4503d9d60159aecc074/server/zanata-frontend/src/app/editor/components/TransUnitTranslationFooter.js#L94)):*
> ```JavaScript
> 
>     const saveCallback = (event) => {
>       if (status === 'rejected') {
> ```

We should use const for all the status type

---

*[server/zanata-frontend/src/app/editor/reducers/review-trans-reducer.js, line 9 at r8](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4cgQPOd-TmA7zKgAGf:-L4cgQPOd-TmA7zKgAGg:bnmuwhf) ([raw file](https://github.com/zanata/zanata-platform/blob/2508c8e86875b1baf135e4503d9d60159aecc074/server/zanata-frontend/src/app/editor/reducers/review-trans-reducer.js#L9)):*
> ```JavaScript
> 
> const defaultState = {
>   notification: null,
> ```

I think we use 'undefined' in js instead of null

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/688)*
<!-- Sent from Reviewable.io -->

aeng commented:
Reviewed 1 of 14 files at r4.
Review status: 24 of 26 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/688)*
<!-- Sent from Reviewable.io -->

efloden commented:
Review status: 20 of 32 files reviewed at latest revision, 3 unresolved discussions.

---

*[server/zanata-frontend/src/app/editor/components/TransUnitTranslationFooter.js, line 94 at r8](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4cgeZKi5lIm7U4WWPr:-L4cua_KgDGglfSTvuco:b6yjyq) ([raw file](https://github.com/zanata/zanata-platform/blob/2508c8e86875b1baf135e4503d9d60159aecc074/server/zanata-frontend/src/app/editor/components/TransUnitTranslationFooter.js#L94)):*
<details><summary><i>Previously, aeng (Alex Eng) wrote…</i></summary><blockquote>

We should use const for all the status type
</blockquote></details>

Imported and using the STATUS_REJECTED const from the phrase utils

---

*[server/zanata-frontend/src/app/editor/components/TransUnit/index.js, line 141 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4Ew47gpj5hPB3ClMpe:-L4cv0GZVknEsEKqt5iq:bqiqvff) ([raw file](https://github.com/zanata/zanata-platform/blob/1e68d193698c94ad1b0193e533db3fed685a32ad/server/zanata-frontend/src/app/editor/components/TransUnit/index.js#L141)):*
<details><summary><i>Previously, huangp (Patrick Huang) wrote…</i></summary><blockquote>

oh I just realize we are rendering the modal component on every TransUnit. That's unnecessary. We should just render one RejectTranslationModal on the editor page (e.g. in mainContent). We can get selected transUnitID and current localeID from state.
</blockquote></details>

Moved the component to render once in MainContent

---

*[server/zanata-frontend/src/app/editor/reducers/review-trans-reducer.js, line 9 at r8](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4cgQPOd-TmA7zKgAGf:-L4culyDwN0E-VwPMOAi:b-px8bsm) ([raw file](https://github.com/zanata/zanata-platform/blob/2508c8e86875b1baf135e4503d9d60159aecc074/server/zanata-frontend/src/app/editor/reducers/review-trans-reducer.js#L9)):*
<details><summary><i>Previously, aeng (Alex Eng) wrote…</i></summary><blockquote>

I think we use 'undefined' in js instead of null
</blockquote></details>

Refactored this in a few additional places, updated the tests to check for undefined as well

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/688)*
<!-- Sent from Reviewable.io -->

aeng commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 12 of 12 files at r9.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-:-L4d6QKSS5aQ2rRF2j8f:bnfp4nl)*
<!-- Sent from Reviewable.io -->

spathare21 commented:
@here a review remains or you forgot to add reviewed : )

efloden commented:
@spathare21 just noticed a bug sorry, working on that before QA

spathare21 commented:
@efloden 

- even we cancel the reject translation pop-up it shows translation rejected. 

- Need work text not wrapped correctly when Rejected is select in dropdown

-- When Translated selected 
![screenshot from 2018-02-07 13-38-45](https://user-images.githubusercontent.com/23230884/35905262-4eea0d36-0bde-11e8-92af-88d7f7ba8606.png)

-- When Rejected selected
![screenshot from 2018-02-07 12-46-32](https://user-images.githubusercontent.com/23230884/35905267-522f5780-0bde-11e8-945e-53d4c1314b3b.png)
 
- Criteria drop-down should be short or arrow of drop-down should be at end.

![screenshot from 2018-02-07 13-41-45](https://user-images.githubusercontent.com/23230884/35905508-54e37456-0bdf-11e8-80d8-cd316d125d70.png)

- Count display in editor on top left corner is not changing for  rejected, translated, needworks without refresh

![screenshot from 2018-02-07 13-42-26](https://user-images.githubusercontent.com/23230884/35905560-88fe72e0-0bdf-11e8-8c8b-d26e3c2267e0.png)

- Unable to open reject translation pop-up with Ctr + Shift + S R

- In firefox browser if we press Ctrl + Shift + S R it shows following screens

![screenshot from 2018-02-07 14-02-50](https://user-images.githubusercontent.com/23230884/35906333-ffec1eb4-0be1-11e8-9d6d-6d0adeee0bc4.png)

![screenshot from 2018-02-07 14-03-10](https://user-images.githubusercontent.com/23230884/35906337-018f216c-0be2-11e8-9cb1-5a18c4033bd6.png)


- There should be a character limit in comment field of reject translation

efloden commented:
@spathare21 tried to address all of these issues except the count display, which I will try to solve in a separate PR as it already has an associated ticket ([ZNTA-1669](https://zanata.atlassian.net/browse/ZNTA-1669))
I can't reproduce the key combination bug in firefox, it appears to be complaining about chrome selenium?
CTRL + SHIFT + S should open the dropdown with the key view:
![keys](https://user-images.githubusercontent.com/7971187/35956272-51a90d94-0ce0-11e8-8555-52132e49ab6f.png)
The R key was not registering because the page was focusing on the <body> element after the key combination, I've resolved this by re-focusing the dropdown when the key combination is entered.

spathare21 commented:
@efloden thanks will test it after reviewed

aeng commented:
Reviewed 2 of 5 files at r11, 2 of 2 files at r12, 10 of 11 files at r13, 1 of 1 files at r14.
Review status: 34 of 35 files reviewed at latest revision, 2 unresolved discussions.

---

*[server/services/src/main/java/org/zanata/rest/editor/dto/ReviewData.java, line 29 at r14](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4rxQe82KGamKlY3-sy:-L4rxQe82KGamKlY3-sz:b-7obgss) ([raw file](https://github.com/zanata/zanata-platform/blob/f569ca4a6b805ceeace4e92a27389ec9e2697428/server/services/src/main/java/org/zanata/rest/editor/dto/ReviewData.java#L29)):*
> ```Java
> 
>     @JsonProperty("comment")
>     private String comment;
> ```

Is there any length limit on this field? You might want to use the annotation @Size()  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/688)*
<!-- Sent from Reviewable.io -->

efloden commented:
Review status: 34 of 35 files reviewed at latest revision, 2 unresolved discussions.

---

*[server/services/src/main/java/org/zanata/rest/editor/dto/ReviewData.java, line 29 at r14](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L4rxQe82KGamKlY3-sy:-L4s1_0PGZ9yxLDNnjAv:b-x6g6i0) ([raw file](https://github.com/zanata/zanata-platform/blob/f569ca4a6b805ceeace4e92a27389ec9e2697428/server/services/src/main/java/org/zanata/rest/editor/dto/ReviewData.java#L29)):*
<details><summary><i>Previously, aeng (Alex Eng) wrote…</i></summary><blockquote>

Is there any length limit on this field? You might want to use the annotation @Size()  
</blockquote></details>

Limiting it here with @Size to match the 500 char limit in the editor

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/688)*
<!-- Sent from Reviewable.io -->

aeng commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r15.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-:-L4s39vmpJ6EVFnvUe3c:bnfp4nl)*
<!-- Sent from Reviewable.io -->

spathare21 commented:
@efloden 
- Save string as translated taking more time. 


![screenshot from 2018-02-09 11-42-58](https://user-images.githubusercontent.com/23230884/36014231-b003fb48-0d8e-11e8-9e52-c090f9b771a5.png)


- Even after selecting rejected and entering comment translated string is not rejected

-  Ctr + Shift + S opens the dropwdown with key view but still pressing R dosen''t open rejection pop-up

- Comment field limits the character but not showing how many characters are remaning/limited. will be good to implement character count while entering comment.

![screenshot from 2018-02-09 11-49-23](https://user-images.githubusercontent.com/23230884/36014354-4d4d8716-0d8f-11e8-91c6-c9a68b3569a0.png)


- After updating my firefox issue mentioned in prevoius comment is not occuring.

spathare21 commented:
@efloden thanks for fixing the above issues

Key commands and Comment filed looks fine for me. 

but still looks like the save translation, need works, Approved taking long time in Alpha editor.
and still not able to save as rejected, haven't observed any relevant logs in console. but may be this is usefull 

`06:03:20,527 INFO  [javax.enterprise.resource.webcontainer.jsf.renderkit] (default task-14) WARNING: FacesMessage(s) have been enqueued, but may not have been displayed.
sourceId=null[severity=(ERROR 2), summary=(One or more resources have the target of 'head', but no 'head' component has been defined within the view.), detail=(One or more resources have the target of 'head', but no 'head' component has been defined within the view.)]
06:04:00,747 INFO  [javax.enterprise.resource.webcontainer.jsf.renderkit] (default task-36) WARNING: FacesMessage(s) have been enqueued, but may not have been displayed.
sourceId=null[severity=(ERROR 2), summary=(One or more resources have the target of 'head', but no 'head' component has been defined within the view.), detail=(One or more resources have the target of 'head', but no 'head' component has been defined within the view.)]
06:04:15,324 INFO  [javax.enterprise.resource.webcontainer.jsf.renderkit] (default task-58) WARNING: FacesMessage(s) have been enqueued, but may not have been displayed.
sourceId=null[severity=(ERROR 2), summary=(One or more resources have the target of 'head', but no 'head' component has been defined within the view.), detail=(One or more resources have the target of 'head', but no 'head' component has been defined within the view.)]
`

![screenshot from 2018-02-12 11-37-36](https://user-images.githubusercontent.com/23230884/36085096-2c0bbfc6-0fe9-11e8-91b6-88207fc0230a.png)

efloden commented:
@spathare21 some fixes/updates:
Fixed a bug where the criteria object was sent as the criteriaId after selecting a criteria.
It is now possible to reject a translation without assigning a criteria (as in the gwt editor).
If there are no criteria created then the options for criteria and priority are now hidden in the modal.
Validation checking is in place to ensure there is either a criteria selected or comment written for a rejection.

spathare21 commented:
@efloden build failed

spathare21 commented:
@efloden even though there is no review criteria it shows select criteria and priority 
other looks fine.
![screenshot from 2018-02-13 17-52-02](https://user-images.githubusercontent.com/23230884/36149679-939ffce0-10b8-11e8-969f-82b85aa608c5.png)

efloden commented:
Review status: 32 of 35 files reviewed at latest revision, 4 unresolved discussions.

---

*[server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js, line 94 at r19](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L5Gg2wIF-pMck2AbgFL:-L5GjdHD_MnkOaADWYEF:b-qr3cmj) ([raw file](https://github.com/zanata/zanata-platform/blob/042f9b71da19338e1b08f13446177e3dd1534499/server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js#L94)):*
<details><summary><i>Previously, aeng (Alex Eng) wrote…</i></summary><blockquote>

Is review.reviewComment a string? Think we can use lodash to check isBlank or our StringUtil method.
</blockquote></details>

Yes, changed the check for reviewComment to use the lodash isEmpty method

---

*[server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js, line 110 at r19](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L5GfvEIiwYc6vNURl7a:-L5Gh4qoimoTJ6zMqs4a:bodxz6n) ([raw file](https://github.com/zanata/zanata-platform/blob/042f9b71da19338e1b08f13446177e3dd1534499/server/zanata-frontend/src/app/editor/containers/RejectTranslationModal/index.js#L110)):*
<details><summary><i>Previously, aeng (Alex Eng) wrote…</i></summary><blockquote>

Is criteria a string? Think we can use lodash to check isBlank or our StringUtil method.
</blockquote></details>

criteria is an array of criteria, renamed this to criteriaList for clarity and used the lodash isEmpty method here as well

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/688)*
<!-- Sent from Reviewable.io -->

aeng commented:
Reviewed 2 of 2 files at r20.
Review status: 34 of 35 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/688)*
<!-- Sent from Reviewable.io -->

efloden commented:
Review status: 34 of 35 files reviewed at latest revision, 2 unresolved discussions.

---

*[server/services/src/main/java/org/zanata/rest/editor/service/TranslationReviewService.java, line 88 at r19](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-L5Geb7EdPkR-A9fCjNg:-L5GoYHxkmCqEz__qltc:b-z309e3) ([raw file](https://github.com/zanata/zanata-platform/blob/042f9b71da19338e1b08f13446177e3dd1534499/server/services/src/main/java/org/zanata/rest/editor/service/TranslationReviewService.java#L88)):*
<details><summary><i>Previously, aeng (Alex Eng) wrote…</i></summary><blockquote>

I think we can refactor these 2 lines for more readable
```
ReviewCriteria criteria = null;
if (data.getReviewCriteriaId != null) {
    ReviewCriterionId reviewId = new ReviewCriterionId(data.getReviewCriteriaId())
    criteria = dao.findBy(reviewId.getId())
}
```
</blockquote></details>

refactored this to be more legible

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/688)*
<!-- Sent from Reviewable.io -->

aeng commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r21.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/688#-:-L5Gq-k6cJCPn3CBMUnj:bnfp4nl)*
<!-- Sent from Reviewable.io -->
